File: node/arduino.js
const SerialPort = require('serialport');
const StringDecoder = new (require('string_decoder').StringDecoder)('utf8');

module.exports = class Arduino {
    constructor() {
        this.device = null
        // to filter the arduino device in the list
        this.productId = '7523'
        this.path
        this.baudRate = 9600
        this.lastResponse = {}
    }

    init() {
        return new Promise((resolve, reject) => {
            // list all serial port and take the first as the arduino
            SerialPort.list().then((ports) => {
                this.path = ports.filter(item => item.productId === this.productId)[0].path

                console.log('ARDUINO: Arduino plugged on ' + this.path)
                this.device = SerialPort(
                    this.path,
                    {
                        baudRate: this.baudRate
                    }
                )

                this.device.on('open', () => {
                    // if (err) {
                    //     console.log("ARDUINO: can't init serialport connection")
                    //     console.log(err)
                    //     return reject()
                    // }
                    setTimeout(() => {
                        return resolve()
                    }, 900)
                })
            }).catch((err) => {
                if (err !== null) {
                    console.log('ARDUINO: Error while getting the list of serialports')
                    console.log(err)
                    return reject()
                }
            })
        })
    }

    parseResponse(data) {
        let decodedData = StringDecoder.write(data)
        let responseType = decodedData.split(': ')[0]
        console.log('decoded Data', decodedData)
        let payload = decodedData.split(': ')[1].replace('\r\n', '')
        return {
            responseType,
            payload
        }
    }

    sendCommand(name, params = [], expectResponse = false) {
        // console.log(name)
        return new Promise((resolve, reject) => {
            let toSend = name.toUpperCase()
            if (params.length <= 4) {
                params.forEach(p => {
                    toSend += '#' + p
                })
            }
            toSend += '\n'
            if (expectResponse) {
                this.device.removeAllListeners('data')
                this.device.on('data', data => {
                    return resolve(this.parseResponse(data))
                })
            }
            console.log('sent', toSend)
            this.device.write(toSend, (err) => {
                if (err !== undefined) {
                    console.log("ARDUINO: Can't send a command")
                    console.log(err)
                    return reject()
                } else {
                    if (!expectResponse) {
                        resolve()
                    }
                }
            })
        })
    }


    /**
     * Send a ping to known if we can communicate correctly with the arduino
     */
    isAlive() {
        return new Promise((resolve, reject) => {
            this.sendCommand('PING', [], true).then(response => {
                console.log(response.payload)
                if (response.payload !== 'pong!' || response.responseType !== 'L') {
                    return reject()
                } else {
                    return resolve()
                }
            })
        })
    }
}

File: python/src/Arduino.py
import serial
import time
from .ThreadHelper import Thread

class Arduino:
  name = ''

  def __init__(self, path, baudrate = 9600):
    self.logger = self.container.get('logger').get('Arduino')
    self.serial = serial.Serial(
      port = path,
      baudrate = baudrate,
      parity = serial.PARITY_NONE,
      stopbits = serial.STOPBITS_ONE,
      bytesize = serial.EIGHTBITS,
      timeout = 2
    )
    
  def getName(self):
    return self.name

  # def watchPort(self):
  #   while True:
  #     line = self.serial.read(1)
  #     line = line.decode('utf-8')
  #     print(line)

  # def startReadThread(self):
  #   self.readThread = Thread(target=self.watchPort)
  #   self.readThread.start()

  def readLine(self):
    l = self.serial.readline()
    return l.decode('utf-8')

  def init(self):
    self.logger.info("Waiting for arduino serial connexion...")
    line = ''
    while "Pong!" not in line:
      line = self.sendCommand(
        name = "PING",
        expectResponse = True,
        noFilter = True
      )
    self.logger.info("Initialized!")
    
  def identify(self):
    line = self.sendCommand(
      name = "ID",
      expectResponse = True
    )
    self.name = line.split("ID:")[1][0:-1]
    return self.name
  
  def setName(self, name):
    self.name = name

  def sendCommand(self, **options):
    paramsString = ''
    if 'params' in options:
      for param in options['params']:
        if isinstance(param, int):
          paramsString += '#' + str(param)
        else: 
          paramsString += '#' + param
    command = options['name'] + paramsString
    #print('command sent:', command , 'on device', self.name)

    name = self.name
    if len(name) > 0:
      name = name + ':'

    self.logger.debug(name + ' --> ' + command)
    self.serial.write(str.encode(command))
    if 'expectResponse' in options and options['expectResponse']:
      # print(self.serial.is_open)
      # self.serial.read(self.serial.in_waiting)
      # print(self.serial.in_waiting)
      if 'noFilter' in options and options['noFilter']:
        line = self.readLine()
      else:
        line = ''
        while len(line) < 2:
          line = self.readLine()
          #print(line)
      #print('response expected:', line)
      
      self.logger.debug(name + ' <-- ' + line.replace('\n', ''))
      return line
        
